src/utilities/HiFiMath.ts

```diff
@@ -0,0 +1,460 @@
+/**
+ * This Module contains a minimal set of 3D math utilities.
+ * @packageDocumentation
+ */
+
+import { HiFiUtilities } from "./HiFiUtilities";
+
+const MIN_NORMALIZABLE_SQUARE_LENGTH = 1.0e-15;
+const RADIANS_TO_DEGREES = 180.0 / Math.PI;
+const DEGREES_TO_RADIANS = Math.PI / 180.0;
+
+/**
+ * A point in 3D space.
+ */
+export class Vector3 {
+    x: number;
+    y: number;
+    z: number;
+
+    constructor({ x = 0, y = 0, z = 0 }: { x?: number, y?: number, z?: number } = {}) {
+        this.x = x;
+        this.y = y;
+        this.z = z;
+    }
+
+    negate() {
+        this.x = -this.x;
+        this.y = -this.y;
+        this.z = -this.z;
+    }
+
+    length(): number {
+        return Math.sqrt(this.x * this.x + this.y * this.y + this.z * this.z);
+    }
+
+    length2(): number {
+        return this.x * this.x + this.y * this.y + this.z * this.z;
+    }
+
+    static add(a: Vector3, b: Vector3) {
+        return new Vector3({x: a.x + b.x, y: a.y + b.y, z: a.z + b.z});
+    }
+
+    static subtract(a: Vector3, b: Vector3) {
+        return new Vector3({x: a.x + b.x, y: a.y + b.y, z: a.z + b.z});
+    }
+
+    static scale(s: number, b: Vector3) {
+        return new Vector3({x: s * b.x, y: s * b.y, z: s * b.z});
+    }
+
+    static dot(a: Vector3, b: Vector3): number {
+        return a.x * b.x + a.y * b.y + a.z * b.z;
+    }
+
+    static cross(a: Vector3, b: Vector3): Vector3 {
+        return new Vector3({x: a.y * b.z - a.z * b.y, y: a.z * b.x - a.x * b.z, z: a.x * b.y - a.y * b.x});
+    }
+
+    static distance(a: Vector3, b: Vector3): number {
+        let x = b.x - a.x;
+        let y = b.y - a.y;
+        let z = b.z - a.z;
+        return Math.sqrt(x*x + y*y + z*z);
+    }
+
+    static distance2(a: Vector3, b: Vector3): number {
+        let x = b.x - a.x;
+        let y = b.y - a.y;
+        let z = b.z - a.z;
+        return x*x + y*y + z*z;
+    }
+}
+
+/**
+ * A Quaternion for 3D rotations.
+ */
```

Comment:
A nearly complete `Quaternion` implementation was required, mostly so that we could supply unit-tests to verify the more customer-centric utils are working correctly.